### ISSUE TITLE
Validate results are "same" as old projects

### ISSUE THREAD

**Issue by e-larsson** [Scan-o-Matic/scanomatic-standalone#145]
# 📋 Info

The results produced by the system does not seem to be coherent with the old projects, here is a summary of a previous investigation for the `PayamMartin19Oct16Scanner2` project, it includes the compilation/analysis/feature extraction steps but NOT scanning:

# 2 - Compile Project
With local fixture
## Produces
- `*.project.compilation`
- `*.project.compilation.instructions`

### `*.project.compilation`
- Skipped index 0 image
- Grayscale has slight differences in coordinates and decimals in values
- Orientation marks slightly different
- FixturePlateModel coordinates slightly different

## Errors
`2022-01-28 08:21:55 -- ERROR	**Compile Effector** Could not output analysis to file /somprojects/PayamMartin19Oct16Scanner2/PayamMartin19Oct16Scanner2_0000_30.2968.tiff`

# 3 - Analysis
With dynamic positioning
## Produces
- `grid_plate___*.npy`
- `grid_size___*.npy`
- `time_data.npy`
- `image_*_data.npy`

### `grid_plate___*.npy`
Same shape but slightly different values by single digits

### `image_*_data.npy`
- The file with the last index is missing, due to missing index 0 in compilation probably
- The numbers are different by about factor 2

### `time_data.npy`
- Shape is off by one, first element is missing, due to missing index 0 in compilation probably

# 4 - Feature extraction
Run with overwriting QC
## Produces
- `curves_raw.npy`
- `curves_smooth.npy`
- `normalized_phenotypes.npy`
- `phenotype_params.json`
- `phenotypes.Absolute.plate_*.csv`
- `phenotypes.extraction.instructions`
- `phenotypes_filter.npy`
- `phenotypes_meta_vector_raw.npy`
- `phenotypes_raw.npy`
- `phenotypes_reference_offsets.npy`
- `phenotype_vectors_raw.npy`
- `phenotype_times.npy`

### `curves_raw.npy` and `curves_smooth.npy`
- Shape is off by one, probably same issue as earlier
- Numbers are off by about factor 2

### `phenotypes.Absolute.plate_*.csv`
- 5 more columns in new version
- Numbers are off

### `phenotypes_filter.npy`
- The filters are different on a few (18 out of 1536) positions for one of the plates

### `phenotypes_raw.npy`
- New format so difficult to compare

### `phenotype_vectors_raw.npy`
- Old one is empty

### `phenotype_times.npy`
- First element is missing

# 🏁 DoD

- [x] Investigate the error message in # 2
- [x] Investigate why the image with index 0 is not properly loaded
- [ ] Ensure results are same/similar to old project
- [x] Validate `phenotypes_filters.npy` is same/similar to old one, it should be able to be opened

**Comment by e-larsson** [comment]
Some previous discussion can be found in https://github.com/Scan-o-Matic/scanomatic-standalone/pull/144

**Comment by local-minimum** [comment]
Consider checking the `values` of the grayscale in fixture vs `project.compilation` file in original data and then check the values in the fixture vs the values in a newly compiled `project.compilation`.

**Comment by BengtRydberg** [comment]
Below is a figure showing a comparison between new and old grayscale values for PayamMartin19Oct16Scanner2 project. Differences are found to be within about +- 0.13. Is this an issue or not?

![scanomatic_grayscale_values](https://user-images.githubusercontent.com/48437915/153385173-6b9b303c-5ea9-4849-8d62-c8832e75a23e.png)

**Comment by local-minimum** [comment]
It looks quite good to me, if you want to check if a bit further each array of values is used to fit a polynomial (don't remember exactly but might be a 3rd deg poly) and it's the differences in values for the range `[0, 255]` that really matters.

**Comment by skymandr** [comment]
I don't know if this is an issue, but I feel fairly confident in saying that this can't explain the difference we are seeing.

**Comment by local-minimum** [comment]
me too lets say it's something else that causes the differences then... How does heatmaps of differences between the raw growthcurve values look?

**Comment by BengtRydberg** [comment]
> me too lets say it's something else that causes the differences then... How does heatmaps of differences between the raw growthcurve values look?

What file(s) contain this information?

**Comment by skymandr** [comment]
> me too lets say it's something else that causes the differences then... How does heatmaps of differences between the raw growthcurve values look?

@e-larsson looked a bit at this. The shapes of the curves are basically the same, but the values are different by a factor of approximately two (but the difference was, as I recall, time dependent). The difference is similar to that you get when comparing _e.g._ `analysis` with `analysis_original` and `analysis_no_median`. @local-minimum: Do you know what the difference is between these different analyses?

**Comment by local-minimum** [comment]
no_median probably refers to skipping median filter in curve smoothing, but I doubt that's a primary reason for big changes and if it's as I expect it would only be causing difference on the smooth curves and phenotypes, not on the raw curves.

**Comment by skymandr** [comment]
Our current line of inquiry is to make a new compilation and analysis on a project in the 2.X-branch, and compare that with results from the 3.X branch, to make sure we are comparing the same things.

**Comment by local-minimum** [comment]
Makes sense, if someone has time to look at other project we have it might be worth just trying one named something completely different and see if numbers agree better in that one.

**Comment by BengtRydberg** [comment]
Below is a comparison between the image 213 in the analysis for PayamMartin19Oct16Scanner2.
I have compared three images:

  - python2.7: from joakims run and with a python2.7 docker image
  - python3.9: from my run and with a python3.9 docker image
  - payam: from PayamMartin19Oct16Scanner2/analysis_original

and the percentage difference is reported as 200 * (x - y) / (x + y).

From the figures we see that both data from python2.7 and python3.9 deviates quite significantly from
payams original analysis. But it seems that pixel values in the images from python2.7 and python3.6
agrees closely. There are a few pixels that deviates by ~25 % for the upper right plate.
However, from the bottom figure showing a histogram of the differences we see that most pixels
agree within 0.5 % even for this plate.

So I guess this is good news for us...

![scanomatic_1](https://user-images.githubusercontent.com/48437915/153504622-d8d4bc88-44bb-4532-85ec-2e08fd136852.png)

![scanomatic_2](https://user-images.githubusercontent.com/48437915/153504632-a9d1b446-72f6-4dc5-9114-a37414015137.png)

![scanomatic_3](https://user-images.githubusercontent.com/48437915/153504663-c0f45650-56db-464f-9d7b-028cdb86bac8.png)

![scanomatic_4](https://user-images.githubusercontent.com/48437915/153504676-67491fc4-e7d0-4410-93ab-5d0f11a49ca1.png)

**Comment by local-minimum** [comment]
I agree that it looks like good news for us!

It is weird that the comparison with payam shows a spacial bias, which makes me wonder if that data had been corrected for that bias in some manner.

The differences between 2.7 och 3.9 look benign. Since it is a non-trivial and non-deterministic process it's to be expected that there be some variation. And this is also part of the reason why only ever relative experiments compared to some reference experiment is done.

**Comment by local-minimum** [comment]
It would be nice to also see that the phenotypes and normalized phenotypes (check growth rate) are comparable between 2.7 and 3.9 if it's not too much work.

**Comment by BengtRydberg** [comment]
Same type of figures but for Phenotypes.GenerationTime

![scanomatic_11](https://user-images.githubusercontent.com/48437915/153594268-d3200d7a-f872-482a-9219-124987e88561.png)

![scanomatic_12](https://user-images.githubusercontent.com/48437915/153594276-453922ca-ed18-4dea-b35e-119f8b0d28e9.png)

![scanomatic_14](https://user-images.githubusercontent.com/48437915/153594291-93aeac30-80df-46f5-a000-428a8ebb0834.png)

![scanomatic_13](https://user-images.githubusercontent.com/48437915/153594317-7aef4ca3-c909-4a6c-85c6-6e5b4d38c818.png)

**Comment by local-minimum** [comment]
It's a bit disturbing that there's a positional bias in the comparison between python 2.7 and 3.9. I don't quite understand what might be causing that. That there might be a very minute general shift independent of position on the lower plates seems not troublesome but the top two plates are scary.

**Comment by skymandr** [comment]
> It's a bit disturbing that there's a positional bias in the comparison between python 2.7 and 3.9. I don't quite understand what might be causing that. That there might be a very minute general shift independent of position on the lower plates seems not troublesome but the top two plates are scary.

I think it is only the top left plate that is scary. I don't see the same spatial tendency in the top-right one, though the missing data in that plate makes it difficult to say for sure. I would suggest that a good starting point i code inspection: look at the `GenerationTime` phenotyper code in 2.7 and 3.8 side by side to rule out obvious things, such as int vs. float division.

**Comment by local-minimum** [comment]
I didn't realize it was missing, but why is it missing with a spatial bias then? 

At any rate this must be investigated before we can consider this task done.

**Comment by skymandr** [comment]
> I didn't realize it was missing, but why is it missing with a spatial bias then?

I don't know, but it's the same in the analyses we have from before, if I recall correctly. It is very evident in QC, when looking at the affected plate.

**Comment by joakimmoller** [comment]
Next agreed step is to narrow down were the differences starts to occur. "Start process in 2.7 continue in in python3"...

I.e. run analysis in scanomatic 2.2 export the data and continue with feature extraction in scanomatic 3.

**Comment by skymandr** [comment]
Some more preliminary results, where I've tried to replicate the spatial bias seen for the Generation Time phenotype (GT) above.
![image](https://user-images.githubusercontent.com/358614/155118025-0998927d-62f5-47d5-8d2c-0497675a9457.png)

As can be seen, I cannot reproduce the problem, though the growth yields seem the same as before, as do the ranges of differences in GT between the two versions. It should be noted that I am not 100 % sure the settings in 2 and 3 are exactly the same, but if they are not, then that in it self is an indicator of robustness, rather than the opposite. Still, caution is advised.

A possible explanation can be found in the grayscale differences further up: inspecting the positions that show the spatial bias in QC, I could see that their GT is calculated in roughly the interval where the grayscale is jumpy, whereas the positions lacking bias have their growth phase later. This could explain how the bias came to be.

What is unknown is why the grayscale was jumpy previously but not now, but that the grayscale algorithm can be jumpy is a known problem, so maybe just bad luck? Another explanation could be different fixtures used in for the 2 and 3 analyses in the previous investigation, but that is speculation.

**Comment by local-minimum** [comment]
Sounds good, could you also just make a comparison of the normalized phenotypes and then we might call it a day and say validation is done.

**Comment by skymandr** [comment]
Results from the more careful study are in agreement:

# Raw Yield:
![image](https://user-images.githubusercontent.com/358614/155164421-0424689b-b7ec-4c9e-9587-c85b6cff2c66.png)

Normalized Yield:
![image](https://user-images.githubusercontent.com/358614/155165414-29e6c75e-d4d6-4a2f-a947-e3d8b5f30859.png)

# Raw Growth Time:
![image](https://user-images.githubusercontent.com/358614/155163898-97ac8a86-0f35-460b-9633-79a0efa4baad.png)

# Normalized Growth Time:
![image](https://user-images.githubusercontent.com/358614/155164978-3e4416a0-198b-4158-9fa4-2aceba5a0078.png)

**Comment by local-minimum** [comment]
That looks scary

**Comment by local-minimum** [comment]
> # Normalized Yield:
> ### (scale clipped to [-100, 100] in comparison)
> 
> ![image](https://user-images.githubusercontent.com/358614/155167189-d177c0f0-4929-4857-9ddb-be673852f483.png)


What is the figure 3 here, it truly looks like something went wrong with the clear positional frame.

**Comment by skymandr** [comment]
This is what QC looks like for normalized Yield:

# SoM 2
![image](https://user-images.githubusercontent.com/358614/155272641-b211d166-9997-4d0b-a000-42902003004a.png)

# SoM 3
![image](https://user-images.githubusercontent.com/358614/155272694-5df93daa-e4c6-40c6-ad6a-d7fc4cfba964.png)

**Comment by local-minimum** [comment]
True, percentage isn't as reasonable here. One would realistically take the median/mean reference position value and add it back to the normalized values before doing a percent comparison. And I don't see anything truly troubling here when just subtracting. I think we can say that remaining differences are so minor that they are probably not caused by any serious and new bug introduced. Rather that we are seeing the limits of SoM in its current implementation and since it's not deterministic it is expected. 

I'm happy with these results if you are happy @skymandr

**Comment by skymandr** [comment]
> I'm happy with these results if you are happy @skymandr

I'm happy, but do we need an issue for the weird patterns in yield after normalization?

**Comment by skymandr** [comment]
The spatial normalization problem is documented in issue #164.
